src/ir_def/check_ir.ml

```diff
@@ -1019,9 +1018,9 @@ and check_pat env pat : val_env =
     let ve2 = check_pat env pat2 in
     t <: pat1.note;
     t <: pat2.note;
-    check env pat.at (T.Env.is_empty ve1 && T.Env.is_empty ve2)
-      "variables are not allowed in pattern alternatives";
-    T.Env.empty
+    let common i1 i2 =
+      { typ = T.lub i1.typ i2.typ; loc_known = false; const = i1.const && i2.const } in
+    T.Env.merge (fun _ -> Lib.Option.map2 common) ve1 ve2
```

Comment:
I am super insecure about this implementation.

Reply:
Especially the const bit? What if they are different values, but both const?

Reply:
Hmmm, `check_pat` will always have `const = false`, but `gather_pat` needs a closer look as it receives `const` as a passed-in flag. `AltP` there is not symmetrical currently, that needs to be revised too. But if a `LetD` has a `const` RHS, than either alternative will bind to a constant, so the `&&`-connective will give the right result.

Reply:
this should be fixed in 0a4d4ab49

Reply:
We could factor out the `T.Env.merge (fun _ -> Lib.Option.map2 common)` parts to a top-level `merge_val_envs`.

Reply:
Yes, I think I am happy.